### PR TITLE
BUG: DataFrameGroupBy.transform with axis=1 fails (#36308)

### LINKED ISSUE #36308: BUG: DataFrameGroupBy.transform with axis=1 fails
The following seems to happen on all transform groupby kernels:

```
df = DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
df.groupby([1, 1], axis=1).transform("shift")
```

results in `ValueError: Length mismatch: Expected axis has 2 elements, new values have 3 elements`.

I think the issue is in `pandas.core.groupby.generic._wrap_transformed_output`; this method does not take into account `self.axis` when wrapping the output. All that needs to happen is `result = result.T` and to use the index labels rather than the column labels for `columns` there.

### PR DESCRIPTION
**Pull request by ethanator**
- [x] closes #36308 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Updated `pandas.core.groupby.generic._wrap_transformed_output` as suggested by @rhshadrach but also had to modify `pandas.core.groupby.groupby.pct_change` to get 2 tests passed where the operation passed to `transform` is `pct_change`.

All tests passed except for `pandas/tests/io/test_clipboard.py` and `pandas/tests/io/test_parquet.py` that are already broken on the latest `master` as of September 14, 2020 01:15 UTC.

### CODE REVIEW

**Comment by dsaxton** [comment]
Thanks @Ethanator, can you also add a test and bug fix note in the v1.1.3 whatsnew?

**Review by jreback — CHANGES_REQUESTED** [review]
put this on 1.2, pls add a test for shift and for pct_change which was modified as well.

**Comment by rhshadrach** [comment]
@Ethanator Are you still interested in working on this?

**Comment by Ethanator** [comment]
> @Ethanator Are you still interested in working on this?

Yes. I got distracted but I'm adding tests now to wrap this up.

**Comment by rhshadrach** [comment]
@Ethanator I'd like to get this in 1.2, would it be okay if I finished it up?

**Comment by jbrockmendel** [comment]
> I'd like to get this in 1.2, would it be okay if I finished it up?

yes, if you can address comments about tests and merge master this looked almost ready

**Comment by rhshadrach** [comment]
Both `shift` and `pct_change` are included in the transformation functions in the added test.

**Comment by rhshadrach** [comment]
Test fails on Windows 32-bit because DataFrame gets constructed with 32-bit ints but doing `df["b"] = df["b"].astype(int)` results in 64-bit ints. Not sure what a good way to do this is; maybe specify `dtype=int` in the DataFrame construction?

**Comment by jbrockmendel** [comment]
Well which is the "correct" answer?  if you want to get int32 when on a 32bit machine you can do

```
expected["b"] = expected["b"].astype(np.intp)
```

**Comment by rhshadrach** [comment]
Thanks @jbrockmendel, that's exactly what I was asking but it turns out I had it backwards. The result in the test was 64-bit, but using `astype(int)` on expected (to undo the coercion to float in the expected computation) resulted in 32-bit. I've updated the test to use `astype("int64")`.

However, I'm not sure if this is really the correct behavior. Should the result be coming back as 64-bit here as is currently the case?

**Comment by jreback** [comment]
> Thanks @jbrockmendel, that's exactly what I was asking but it turns out I had it backwards. The result in the test was 64-bit, but using `astype(int)` on expected (to undo the coercion to float in the expected computation) resulted in 32-bit. I've updated the test to use `astype("int64")`.
> 
> However, I'm not sure if this is really the correct behavior. Should the result be coming back as 64-bit here as is currently the case?

your inputs are all int64 (on windows as well), so it IS suprising that you are getting an intp back (32-bit on windows). though i suppose somewhere we are constructing a return array for the ints that uses intp. I would say ok for now, but let's open an issue to dig depper on this.

@rhshadrach ok on this otherwise? (lgtm.).

**Comment by rhshadrach** [comment]
> your inputs are all int64 (on windows as well), so it IS suprising that you are getting an intp back (32-bit on windows).

`result` in the test is coming back as int64; so I think this is then okay. Unless you're surprised that `astype(int)` is giving back 32-bit ints on 32-bit Windows? In the call to `astype(int)`, the inputs are floats.

Aside from this, the PR is good on my end.

File: pandas/tests/groupby/transform/test_transform.py [hunk 1]
```diff
@@ -158,7 +158,25 @@ def test_transform_broadcast(tsframe, ts):
...
+        pytest.xfail("tshift is deprecated")
+    args = ("ffill",) if transformation_func == "fillna" else tuple()
+
+    df = DataFrame({"a": [1, 2], "b": [3, 4], "c": [5, 6]}, index=["x", "y"])
+    result = df.groupby([0, 0, 1], axis=1).transform(transformation_func, *args)
+    expected = df.T.groupby([0, 0, 1]).transform(transformation_func, *args).T
+
+    if transformation_func == "diff":
+        # Result contains nans, so transpose coerces to float
+        expected["b"] = expected["b"].astype("int64")
```
Comment (jreback) [inline pandas/tests/groupby/transform/test_transform.py:173]: oh i c you are doing this because we want to compare to ints, got it ,then this is fine.

**Comment by jreback** [comment]
thanks @Ethanator

**Pull request merged**